Sources/Testing/Parameterization/CustomTestArgument.swift

```diff
@@ -0,0 +1,151 @@
+// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+//
+
+#if canImport(Foundation)
+private import Foundation
+#endif
+
+/// A protocol for customizing how arguments passed to parameterized tests are
+/// encoded, which is used to match against when running specific arguments.
+@_spi(ExperimentalParameterizedTesting)
+public protocol CustomTestArgumentEncodable: Sendable {
+  /// Encode this test argument, using the provided context.
+  ///
+  /// - Parameters:
+  ///   - context: Context about this argument which may be useful in encoding
+  ///     a representation of it which is unique to a specific usage of it.
+  ///
+  /// - Throws: Any error encountered during encoding.
+  ///
+  /// The encoded form of a test argument should be stable and unique to allow
+  /// re-running specific test cases of a parameterized test function. For
+  /// optimal performance, large values which are not necessary to uniquely
+  /// identify the test argument later should be omitted. Values encoded do not
+  /// need to be human-readable.
+  ///
+  /// By default, the testing library checks whether a test argument conforms to
+  /// `Encodable` and encodes it using `encode(to:)` if it does. This is
+  /// sufficient for many types, but for some types the `Encodable`-provided
+  /// representation may contain large data payloads that cause poor
+  /// performance, are not stable and unique, or are otherwise deemed unsuitable
+  /// for testing. If the type of the argument does not conform to `Encodable`
+  /// but it does conform to `Identifiable` and its associated `ID` type
+  /// conforms to `Encodable`, the value of calling its `id` property is used as
+  /// the encoded representation.
+  ///
+  /// It is possible that neither of the approaches for encoded representation
+  /// described above are sufficient. If the type of the argument is made to
+  /// conform to ``CustomTestArgumentEncodable``, then the encoded
+  /// representation formed by calling this method is used.
+  func encodeTestArgument(to encoder: any Encoder, in context: Test.Case.Argument.Context) throws
+}
+
+extension Test.Case.Argument.ID {
+  /// Initialize this instance with an ID for the specified test argument.
+  ///
+  /// - Parameters:
+  ///   - value: The value of a test argument for which to get an ID.
+  ///   - context: The context in which the argument was passed.
+  ///
+  /// - Returns: `nil` if an ID cannot be formed from the specified test
+  ///   argument value.
+  ///
+  /// - Throws: Any error encountered while attempting to encode `value`.
+  ///
+  /// This function is not part of the public interface of the testing library.
+  ///
+  /// ## See Also
+  ///
+  /// - ``CustomTestArgumentEncodable``
+  init?(identifying value: some Sendable, in context: Test.Case.Argument.Context) throws {
+#if canImport(Foundation)
+    guard Configuration.current?.isTestArgumentEncodingEnabled ?? false else {
```

Comment:
`Configuration` will be `nil` on detached tasks and non-Swift-owned threads.

Reply:
Understood, I don't think that will be a practical concern in the contexts we care about here

Reply:
I have adjusted the code further to change where we make this determination so any problems here should be even less likely now, and match how `Event` handles deciding which `Configuration` to use